hbase-common/src/main/java/org/apache/hadoop/hbase/io/crypto/tls/HBaseHostnameVerifier.java

```diff
@@ -0,0 +1,274 @@
+        + "common name of the certificate subject: " + cn);
+    }
+  }
+
+  private static boolean matchIdentity(final String host, final String identity,
+    final boolean strict) {
+    // RFC 2818, 3.1. Server Identity
+    // "...Names may contain the wildcard
+    // character * which is considered to match any single domain name
+    // component or component fragment..."
+    // Based on this statement presuming only singular wildcard is legal
+    final int asteriskIdx = identity.indexOf('*');
+    if (asteriskIdx != -1) {
+      final String prefix = identity.substring(0, asteriskIdx);
+      final String suffix = identity.substring(asteriskIdx + 1);
+      if (!prefix.isEmpty() && !host.startsWith(prefix)) {
+        return false;
+      }
+      if (!suffix.isEmpty() && !host.endsWith(suffix)) {
+        return false;
+      }
+      // Additional sanity checks on content selected by wildcard can be done here
+      if (strict) {
+        final String remainder = host.substring(prefix.length(), host.length() - suffix.length());
+        return !remainder.contains(".");
+      }
+      return true;
+    }
+    return host.equalsIgnoreCase(identity);
+  }
+
+  private static boolean matchIdentityStrict(final String host, final String identity) {
+    return matchIdentity(host, identity, true);
+  }
+
+  private static String extractCN(final String subjectPrincipal) throws SSLException {
+    if (subjectPrincipal == null) {
+      return null;
+    }
+    try {
+      final LdapName subjectDN = new LdapName(subjectPrincipal);
+      final List<Rdn> rdns = subjectDN.getRdns();
+      for (int i = rdns.size() - 1; i >= 0; i--) {
+        final Rdn rds = rdns.get(i);
+        final Attributes attributes = rds.toAttributes();
+        final Attribute cn = attributes.get("cn");
+        if (cn != null) {
+          try {
+            final Object value = cn.get();
+            if (value != null) {
+              return value.toString();
+            }
+          } catch (final NoSuchElementException ignore) {
+            // ignore exception
+          } catch (final NamingException ignore) {
+            // ignore exception
+          }
+        }
+      }
+      return null;
+    } catch (final InvalidNameException e) {
+      throw new SSLException(subjectPrincipal + " is not a valid X500 distinguished name");
+    }
+  }
+
+  private static Optional<InetAddress> parseIpAddress(String host) {
+    try {
+      host = host.trim();
+      // Uri strings only work for ipv6 and are wrapped with brackets
+      // Unfortunately InetAddresses can't handle a mixed input, so we
+      // check here and choose which parse method to use.
+      if (host.startsWith("[") && host.endsWith("]")) {
+        return Optional.of(InetAddresses.forUriString(host));
+      } else {
+        return Optional.of(InetAddresses.forString(host));
+      }
+    } catch (Exception e) {
```

Comment:
Is this a normal behavior? I mean, by design, here we could get an invalid ip address? I think we could also pass in a host name here? Then I suppose we should use something like InetAddresses.isInetAddress to test whether this is a valid op address first. Throwing exception is a bit expensive...

Reply:
The reason I did it this way is when I looked at the impl of `isInetAddress` and `forString`, they both do the exact same parsing. `isInetAddress` is `return ipStringToBytes(ipString) != null`, while `forString` calls the same method and throws an exception if null.

I'm not sure of the relative costs of throwing an exception vs having to parse the ip address/hostname twice. I went for trying to avoid parsing twice, but I could do the `isInetAddress` check instead.

Reply:
You're correct that this method will be called for both hostnames and ip addresses. So it may be common to hit the exception case if someone's deployment most often has hostnames passed around instead of ips.

Reply:
For this case, usually it is recommended to add checks instead of throwing exceptions :)

https://www.baeldung.com/java-check-string-number

You can see the micro benchmark in 6.2 section, only 5% percentage of illegal input can cause a big performance impact if we choose to use NumberFormatException to test whether the input is valid.